lib/src/gzip_encoding.dart

```diff
@@ -0,0 +1,230 @@
+  headers[HttpHeaders.contentLengthHeader] = compressedBodyLength.toString();
+  headers['X-Compression-Ratio'] =
+      '$compressionRatio ($compressedBodyLength/$bodyLength)';
+
+  return response.change(headers: headers, body: compressedBody);
+}
+
+/// Returns `true` if [response] can be compressed.
+///
+/// Checks:
+/// - `Content-Encoding`: if already present, can't change it to `gzip`.
+/// - `Content-Type`: checks if [isAlreadyCompressedContentType].
+/// - `Content-Length`: checks if too small for compression (< 512).
+bool canGzipEncodeResponse(Response response) {
+  var headerContentEncoding =
+      response.headers[HttpHeaders.contentEncodingHeader];
+
+  // If the response already defines a `Content-Encoding` header it
+  // won't apply the the `gzip` encoding to preserve the response behavior.
+  if (headerContentEncoding != null && headerContentEncoding.isNotEmpty) {
+    return false;
+  }
+
+  var responseContentLength = response.contentLength;
+
+  // A small body will not benefit from being compressed:
+  if (responseContentLength != null && responseContentLength < 512) {
+    return false;
+  }
+
+  var headerContentType = response.headers[HttpHeaders.contentTypeHeader];
+
+  // Do not compress if the `Content-Type` is an already compressed type:
+  if (headerContentType != null &&
+      isAlreadyCompressedContentType(headerContentType)) {
+    return false;
+  }
+
+  return true;
+}
+
+/// Returns `true` if [contentType] is already compressed.
+///
+/// See [isAlreadyCompressedExtension].
+bool isAlreadyCompressedContentType(String contentType) {
+  contentType = contentType.toLowerCase().trim();
+  if (contentType.isEmpty) return false;
+
+  var idx = contentType.indexOf(';');
+  if (idx >= 0) {
+    contentType = contentType.substring(0, idx).trim();
+  }
+
+  var idx2 = contentType.indexOf('/');
+
+  var type = idx2 >= 0 ? contentType.substring(idx2 + 1) : contentType;
+
+  if (isAlreadyCompressedExtension(type)) {
+    return true;
+  }
+
+  if (type.contains('+')) {
+    var list = type.split('+');
+    var alreadyCompressed =
+        list.where((e) => isAlreadyCompressedExtension(e)).isNotEmpty;
+    return alreadyCompressed;
+  }
+
+  return false;
+}
+
+bool isAlreadyCompressedExtension(String extension) {
+  extension = extension.toLowerCase().trim();
+  if (extension.isEmpty) return false;
+
+  switch (extension) {
+    case 'ico':
+    case 'png':
+    case 'jpg':
+    case 'jpeg':
+
+    case 'avi':
+
+    case 'mp3':
+    case 'mp4':
+    case 'mpeg':
+
+    case 'ogg':
+    case 'ogx':
+
+    case 'weba':
+    case 'webm':
+    case 'webp':
+
+    case 'epub':
+    case 'pdf':
+
+    case 'woff':
+    case 'woff2':
+
+    case 'jar':
+    case 'war':
+
+    case '7z':
+    case 'bz':
+    case 'bz2':
+    case 'gzip':
+    case 'gz':
+    case 'rar':
+    case 'zip':
+      return true;
+    default:
+      return false;
+  }
+}
+
+// Optimized buffer to read the body.
+class _BytesBuffer {
```

Comment:
At this line you are reading all the response data to compress it later:
https://github.com/gmpassos/shelf_gzip/blob/null-safety/lib/src/gzip_encoding.dart#L60

This reads the data without know the size of the content, so needs to update the size of the destiny array, a `Uint8List `.  It's very inefficient if this is loaded into a normal `List<int>` and later to a `Uint8List ` (needed for compression).

Reply:
I think it would be more maintainable if the Stream from read() is converted into a Uint8List using [toList](https://api.flutter.dev/flutter/dart-async/Stream/toList.html) instead.

Reply:
`await response.read().toList()` returns a `List<List<int>>`, then you still needs to convert to a single `Uint8List`, since a generic `List<int>` is impractical, due memory issues and Gzip performance. The only efficient way is to generate a `Uint8List` on the fly, and this is what `_BytesBuffer` does. Unfortunately Dart doesn't have a standard class to do that.

Note that I believe  that Gzip encoding should aways be on in any shelf server, so I'm really careful to give an implementation that can be used with a good performance.

Reply:
... Also the `_BytesBuffer.toUint8List()` will avoid unnecessary copy of internal `Uint8List` if it's already in the correct size.

Reply:
I'm not sure this package should be maintaining a class like this... even if it's faster or uses less memory as you claim.

Reply:
Without that will be impossible to use the Gzip encoding, it will destroy the memory when providing files. Our team for example won't be able to use it at all. This `_BytesBuffer ` class is simple and does only what is need, is not a complex collection class. I understand that clean code is important, but a real world working solution comes first.

Reply:
Ok, please consider factoring this out at some point.